mantidproject/mantid
# Add a setup.py to build mantid-framework and its C++ extensions

### Description of work.

**This is very much a work in progress, so I would love some feedback on whether this is something we'd consider.**

I was motivated to do this while doing the conda build work. At present, to get the python cpp extensions, e.g `_kernel.so` into the correct place on conda we have two options:

1) Build the mantid framework as usual, which will place the mantid folder in `bin/` (  extensions go in `./bin/mantid/kernel/_kernel.so`) which we then move to the site packages directory. This is not good, as it won't register the mantid-framework installation to pip/setuptools.
2) Change the cmake python package commands to call `pip install .`  which will install the mantid python package, placing all the python files in the site packages directory. This adds everything correctly to the conda python path, but misses out the C extensions, which we have to manually install. This means if we pip uninstall mantid for whatever reason, the extensions will remain.



With this PR, provided you are in a conda environment (because of the way I've temporarily fixed the `setup.py` variables) you can do two things:

1. Install the mantid-framework in developer mode with `python setup.py develop`. This will add an egg-link file to your site-packages which points to the mantid source tree, where the extensions have been built and installed in-place. This means I can use this mantid development build in another project (i.e. where I've specified mantid as a project requirement).

2. Install mantid-framework with `python setup.py install` which will build and install the mantid-cpp libs into `install_prefix/bin` and `install_prefix/lib` and build the python extensions and install them into `site_packages/mantid/(kernel/api/geometry)`.

This is inspired by pyarrow (https://github.com/apache/arrow) and dynd-python (https://github.com/libdynd/dynd-python).

**To test:**
1. Use a conda environment
```
conda env create -f ./mantid-developer-*.yml
conda activate mantid-developer
```
2. Run the setup.py file
`python setup.py develop`
3. After this is finished you should be able to do 
`python -c "import mantid.simpleapi"`




*There is no associated issue.*





---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
